scipy/linalg/tests/test_lapack.py

```diff
@@ -2081,6 +2081,140 @@ def test_pttrf_pttrs_NAG(d, e, d_expect, e_expect, b, x_expect):
+        A = vr @ tri @ vr.conj().T
+        # vr is orthogonal
+        z = vr
+
+    else:
+        # d and e are always real per lapack docs.
+        d = generate_random_dtype_array((n,), realtype)
+        e = generate_random_dtype_array((n-1,), realtype)
+
+        # make SPD
+        d = d + 4
+        A = np.diag(d) + np.diag(e, 1) + np.diag(e, -1)
+        z = np.diag(d) + np.diag(e, -1) + np.diag(e, 1)
+    return (d, e, A, z)
+
+
+@pytest.mark.parametrize("dtype,realtype",
+                         zip(DTYPES, REAL_DTYPES + REAL_DTYPES))
+@pytest.mark.parametrize("compz", ["I", "N", "V"])
+def test_pteqr(dtype, realtype, compz):
+    '''
+    Tests the ?pteqr lapack routine for all dtypes and compz parameters.
+    It generates random SPD matrix diagonals d and e, and then confirms
+    correct eigenvalues with scipy.linalg.eig. With applicable compz=I it
+    tests that z can reform A.
+    '''
+    seed(42)
+    atol = 1000*np.finfo(dtype).eps
+    pteqr = get_lapack_funcs(('pteqr'), dtype=dtype)
+
+    n = 10
+
+    d, e, A, z = pteqr_get_d_e_A_z(dtype, realtype, n, compz)
+
+    d_pteqr, e_pteqr, z_pteqr,work, info = pteqr(d=d, e=e, z=z, compz=compz)
+    assert_equal(info, 0, "info = {}, should be 0.".format(info))
+
+    # compare the routine's eigenvalues with scipy.linalg.eig's.
+    assert_allclose(np.sort(eigh(A)[0]), np.sort(d_pteqr), atol=atol)
+
+    if compz == "I" or compz == "V":
+        # verify z_pteqr as orthogonal
+        assert_allclose(z_pteqr @ np.conj(z_pteqr).T, np.identity(n),
+                        atol=atol)
+        # verify that z_pteqr recombines to A
+        assert_allclose(z_pteqr @ np.diag(d_pteqr) @ np.conj(z_pteqr).T,
+                        A, atol=atol)
+
+
+@pytest.mark.parametrize("dtype,realtype",
+                         zip(DTYPES, REAL_DTYPES + REAL_DTYPES))
+@pytest.mark.parametrize("compz", ["I", "N", "V"])
+def test_pteqr_error_non_spd(dtype, realtype, compz):
+    seed(42)
+    pteqr = get_lapack_funcs(('pteqr'), dtype=dtype)
+
+    n = 10
+    d, e, A, z = pteqr_get_d_e_A_z(dtype, realtype, n, compz)
+
+    # test with non-spd matrix
+    d_pteqr, e_pteqr, z_pteqr, work, info = pteqr(d - 4, e, z=z, compz=compz)
+    assert info > 0
+
+
+@pytest.mark.parametrize("dtype,realtype",
+                         zip(DTYPES, REAL_DTYPES + REAL_DTYPES))
+@pytest.mark.parametrize("compz", ["I", "N", "V"])
+def test_pteqr_raise_error_wrong_shape(dtype, realtype, compz):
+    seed(42)
+    pteqr = get_lapack_funcs(('pteqr'), dtype=dtype)
+    n = 10
+    d, e, A, z = pteqr_get_d_e_A_z(dtype, realtype, n, compz)
+    # test with incorrect/incompatible array sizes
+    assert_raises(ValueError, pteqr, d[:-1], e, z=z, compz=compz)
+    assert_raises(ValueError, pteqr, d, e[:-1], z=z, compz=compz)
+    if compz in {"I", "V"}:
+        d_pteqr, e_pteqr, z_pteqr, work, info = pteqr(d, e, z=z[:-1], compz=compz)
+        assert_(info != 0)
```

Comment:
@ilayn Ideally, this would raise a `ValueError`, too, right? How can we implement that?

Reply:
@ilayn even with:
```
<ftype> intent(in,out,copy), depend(n), dimension((compute_z=='N'?1:MAX(1,n)),(compute_z=='N'?0:n)) :: z
```
as you suggested, this still does not raise a `ValueError`; it behaves the same was as it did before.